### DISCUSSION TITLE
Having trouble using playlist_add_items fn for AuthCodeSpotify. Cannot figure out 'items' type.

### DISCUSSION THREAD

**Discussion by ultrex22** [ramsayleung/rspotify#351] [General]
I'm having trouble figuring out what exactly I need to pass to this method. I have looked at the docs and examples and the source code. I think it wants a trait object but I'm unsure how to accomplish this. I have a Vec of TrackId's I was trying to pass in, but it's not happy, lol.

the error:

type mismatch resolving `<&Vec<TrackId> as IntoIterator>::Item == &dyn PlayableId` E0271
 expected trait object `dyn PlayableId`, found struct `TrackId` 
Note: expected reference `&dyn PlayableId` found reference `&TrackId` 
Note: required by a bound in `playlist_add_items`

this is the code causing the issue:

async fn create_playlist(spotify: &AuthCodeSpotify, track_id_list: Vec<TrackId>) {
    let id = UserId::from_id(USER_ID).expect("Error creating UserId");
    let new_playlist = spotify
        .user_playlist_create(&id, "test_list", Some(false), None, Some("test list"))
        .await
        .expect("error creating playlist");
    spotify
        .playlist_add_items(&new_playlist.id, &track_id_list, None)
        .await
        .expect("error adding items");
}

Any suggestions would be welcome. Thank you!

**Comment by marioortizmanero** [comment]
You need to convert `TrackId` to `&dyn PlayableId` with `as`. This is because you can also add episodes to a playlist, so the method has to be "generic". This seems to work:

```rust

async fn create_playlist(spotify: &AuthCodeSpotify, track_id_list: Vec<TrackId>) {
    let id = UserId::from_id(USER_ID).expect("Error creating UserId");
    let new_playlist = spotify
        .user_playlist_create(&id, "test_list", Some(false), None, Some("test list"))
        .await
        .expect("error creating playlist");
    let playable_id_list = track_id_list
        .iter()
        .map(|track| track as &dyn PlayableId);
    spotify
        .playlist_add_items(&new_playlist.id, playable_id_list, None)
        .await
        .expect("error adding items");
}
```

Note that `track_id_list: Vec` didn't specify a generic parameter for `Vec` either, I imagine that was `Vec<TrackId>`.

**Reply by ultrex22** [reply]
Oh thank you so much for replying and the suggestion of code. You are correct it was a vector of track IDs. I see what you've done in the code that makes sense I will give that a try. I think I can also probably do the conversion in the method where I'm creating the list initially but I'm not sure about that maybe I'll give it a try. Well I've been watching reading and studying rust and involved with it basically on a daily basis I have had very little personal time to sit down and actually code until very recent. So I know that a lot of the hurdles I'm running into are just inexperienced / beginner issues but I definitely appreciate the time and the response.

**Reply by marioortizmanero** [reply]
No worries, best of luck with your coding learning!

**Comment by ultrex22** [comment]
I have the basics of the program working now. Thank you!
Another question I should have asked before.  Is there a way to add the year of release to the search method?

**Reply by marioortizmanero** [reply]
It seems like our link to the reference documentation was broken. Here's the new one: https://developer.spotify.com/documentation/web-api/reference/#/operations/search. It seems like the `year` parameter would work?

I've fixed this in #353.

**Comment by ultrex22** [comment]
Good evening. It took me a little bit to figure out what I was looking at 
in this email. Looks like coincidentally to my question you realized one of 
the links was not working properly or incorrect and you put the fix in or 
pull request for it I guess I should say. So then I read everything 
carefully again and I realized I was assuming something false regarding my 
program and the specific API for the search function. I see I can structure 
my query to include the additional filters one of them being year. For some 
reason in my head I was confusing the return type parameter of the search 
method with the string thinking I could only include certain things in the 
request string if that makes any sense at all. So I'm going to take a look 
at the link and try and put myself together a search query with the year. 
Thank you for the time and the effort and the response!

Ron Brilhante
